### PR TITLE
[FLAX] Question Answering Example 

### PR DESCRIPTION
**Pull request by kamalkraj**
# What does this PR do?

Flax Question Answering Example 

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@patrickvonplaten @patil-suraj @sgugger

### CODE REVIEW

File: examples/flax/question-answering/README.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+```
+huggingface-cli repo create bert-qa-squad-test
+```
+
+Next we clone the model repository to add the tokenizer and model files.
+
+```
+git clone https://huggingface.co/<your-username>/bert-qa-squad-test
+```
```
Comment (sgugger) [inline examples/flax/question-answering/README.md:43]: Not linked to this PR per se, but this is not necessary while using the `Repository` API and all Flax examples should be updated to use it, while removing those instructions.

File: examples/flax/question-answering/README.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+```bash
+f1 = 88.50
+exact_match = 81.03
+```
+
+sample Metrics - [tfhub.dev](https://tensorboard.dev/experiment/6gU75Hx8TGCnc6tr4ZgI9Q)
+
+
+### Usage notes
+Note that when contexts are long they may be split into multiple training cases, not all of which may contain
```
Comment (sgugger) [inline examples/flax/question-answering/README.md:90]:
```suggestion

Note that when contexts are long they may be split into multiple training cases, not all of which may contain
```

File: examples/flax/question-answering/utils_qa.py [hunk 3]
```diff
@@ -0,0 +1,427 @@
+# coding=utf-8
```
Comment (sgugger) [inline examples/flax/question-answering/utils_qa.py:1]: If this is a copy from the PyTorch example folder, it should be added [here](https://github.com/huggingface/transformers/blob/04976a32dc555667afa994e8f918cbee88d84a4f/utils/check_copies.py#L31) so the copy is checked by our internal tooling.

**Review by sgugger — APPROVED** [review]
Thanks for adding that example!

**Review by patil-suraj — APPROVED** [review]
Looks really nice! Thanks for adding this example.

File: examples/flax/question-answering/run_qa.py [hunk 4]
```diff
@@ -0,0 +1,893 @@
...
+                This is the output predictions of the model. We can only enter either start or end logits.
+            eval_dataset: Evaluation dataset
+            max_len(:obj:`int`):
+                The maximum length of the output tensor. ( See the model.eval() part for more details )
+        """
+
+        step = 0
+        # create a numpy array and fill it with -100.
+        logits_concat = np.full((len(dataset), max_len), -100, dtype=np.float64)
+        # Now since we have create an array now we will populate it with the outputs gathered using accelerator.gather
```
Comment (patil-suraj) [inline examples/flax/question-answering/run_qa.py:671]: this comment should be adapted slightly, since we don't use `accelerator` here

File: examples/flax/question-answering/utils_qa.py (hunk 3 above)
Comment (kamalkraj) [reply examples/flax/question-answering/utils_qa.py:1]: Done

File: examples/flax/question-answering/run_qa.py [hunk 5]
```diff
@@ -0,0 +1,893 @@
+#!/usr/bin/env python
+# coding=utf-8
+# Copyright 2020 The HuggingFace Team All rights reserved.
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply examples/flax/question-answering/run_qa.py]: done

File: examples/flax/question-answering/run_qa.py [hunk 6]
```diff
@@ -0,0 +1,893 @@
...
+    eval_batch_size = training_args.per_device_eval_batch_size * jax.local_device_count()
+    # endregion
+
+    # region Load model
+    model = FlaxAutoModelForQuestionAnswering.from_pretrained(
+        model_args.model_name_or_path,
+        config=config,
+        cache_dir=model_args.cache_dir,
+        revision=model_args.model_revision,
+        use_auth_token=True if model_args.use_auth_token else None,
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply examples/flax/question-answering/run_qa.py:735]: done.

File: examples/flax/question-answering/run_qa.py [hunk 7]
```diff
@@ -0,0 +1,893 @@
...
+                    write_train_metric(summary_writer, train_metrics, train_time, cur_step)
+
+                epochs.write(
+                    f"Step... ({cur_step}/{total_steps} | Training Loss: {train_metric['loss']}, Learning Rate: {train_metric['learning_rate']})"
+                )
+
+                train_metrics = []
+
+            if (cur_step % training_args.eval_steps == 0 or cur_step % step_per_epoch == 0) and cur_step > 0:
+
```
[the remark this answers is not in the document]
Comment (kamalkraj) [reply examples/flax/question-answering/run_qa.py]: done

File: examples/flax/question-answering/run_qa.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kamalkraj) [reply examples/flax/question-answering/run_qa.py:735]: @patil-suraj 
Why there is no Dynamic loss scaling for dtype `float16` ? 

 https://flax.readthedocs.io/en/latest/_autosummary/flax.optim.DynamicScale.html#flax.optim.DynamicScale

cc @patrickvonplaten

[the remark this answers is not in the document]
Comment (patil-suraj) [reply examples/flax/question-answering/run_qa.py:735]: After the #13098, `dtype` will specify only the `dtype` of computation, so just the forward pass will be in half-precision and grads will still be computed in `fp32` so for this case loss scaling is not required.

True mixed-precision training for flax examples is planned and will include loss scaling, will post more about it soon :) 

Also, note that we don't use flax optimizers here, we are using `optax`.

**Comment by kamalkraj** [comment]
@patil-suraj 
Done changes according to your review.
